### PR TITLE
feat: add stop order snapshot state

### LINKED ISSUE #8470: Stop orders snapshots
Implement snapshot states for stop orders:

```[tasklist]
### Tasks
- [ ] add snapshots for trailing stop orders
- [ ] add snapshots for priced stop orders
- [ ] ensure submission ordering.
```

### PR DESCRIPTION
**Pull request by jeremyletang**
close #8470

### CODE REVIEW

File: protos/sources/vega/events/v1/events.proto [hunk 1]
```diff
@@ -744,6 +751,9 @@ enum BusEventType {
   // Event indicating a spot liquidity provision has been created or updated.
   BUS_EVENT_TYPE_SPOT_LIQUIDITY_PROVISION = 59;
 
+  // Event indicating a spot liquidity provision has been created or updated.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:754]: Is that what this indicates?

File: protos/sources/vega/vega.proto [hunk 2]
```diff
@@ -59,8 +59,8 @@ message StopOrder {
   optional int64 updated_at = 8;
   // ID of the order created once the trigger is hit.
   string order_id = 9;
-  // Order to be submitted once the trigger is hit.
-  // OrderSubmission order_submission = 10;
+  // The party which submitted this stop order
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:62]:
```suggestion
  // The party that submitted this stop order
```

File: protos/sources/vega/events/v1/events.proto [hunk 3]
```diff
@@ -13,6 +13,11 @@ import "vega/vega.proto";
 
 option go_package = "code.vegaprotocol.io/vega/protos/vega/events/v1";
 
+message StopOrderEvent {
```
[the remark this answers is not in the document]
Comment (jeremyletang) [reply protos/sources/vega/events/v1/events.proto:16]: hum,  known of them are commented on this file, so I think it's fine like it is?

File: protos/sources/vega/events/v1/events.proto (hunk 1 above)
Comment (jeremyletang) [reply protos/sources/vega/events/v1/events.proto:754]: correct. will fix.

File: core/execution/stoporders/stop_orders_snapshot_test.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+	buf, err := proto.Marshal(serialized)
+	assert.NoError(t, err)
+
+	deserialized := &v1.StopOrders{}
+	err = proto.Unmarshal(buf, deserialized)
+	assert.NoError(t, err)
+
+	pool2 := stoporders.NewFromProto(log, deserialized)
+	assert.True(t, pool.Equal(pool2))
```
Comment (wwestgarth) [inline core/execution/stoporders/stop_orders_snapshot_test.go:49]: I don't think `pool.Equal()` is doing a strict enough job. If I add this to the end of the test it fails:
```
serialized2 := pool2.ToProto()
buf2, err := proto.Marshal(serialized2)
assert.NoError(t, err)
assert.Equal(t, hex.EncodeToString(buf), hex.EncodeToString(buf2))
```
so I guess the tree isn't being populated in exactly the same order when we restore?

Comment (jeremyletang) [reply core/execution/stoporders/stop_orders_snapshot_test.go:49]: hum interesting. I shall look again then. I was pretty sure I made sure that both trees were containing the same things. and I can find the exact same elements in both. let me look again. thanks for spoting, I should have done that exact thing you did.

**Review by wwestgarth — APPROVED** [review]

**Pull request merged**